File: docs/JAMIA/comparisons/RPython/TM/44893933_case_when_na_type.ts
/**
 * RPython SO#44893933 — Avoiding type conflicts with dplyr::case_when
 * Effect: Crash
 * Bug class: Nullable type
 *
 * R bug: case_when requires all branches to return the same type. Writing
 * `case_when(old == 2 ~ NA)` gives logical NA, while other branches return numeric.
 * This crashes: "must be type double, not logical." Fix: use NA_real_ explicitly.
 *
 * In tidy-ts, a conditional that returns null produces number | null. Downstream
 * operations requiring strict number (like Math.round) catch the nullable type.
 */
import { createDataFrame } from "@tidy-ts/dataframe";

const df = createDataFrame([
  { old: 1 },
  { old: 2 },
  { old: 3 },
]);

// The SO user's intent: case_when(old == 1 ~ 5, old == 2 ~ NA, TRUE ~ old)
// R crashes because NA is logical, not numeric.
// tidy-ts equivalent: returning null from a branch makes the column number | null.
const result = df.mutate({
  new: (r) => {
    if (r.old === 1) return 5;
    if (r.old === 2) return null;
    return r.old;
  },
});

// result.new is number | null. Passing to Math.round (requires number) fails.
// @ts-expect-error — Argument of type 'number | null' is not assignable to parameter of type 'number'
result.mutate({ doubled: (r) => Math.round(r.new) });

File: docs/JAMIA/comparisons/RPython/TM/44893933_case_when_na_type.R
# ID: SO#44893933
# Language: R
# Bug class: Nullable
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: bare NA is logical type; case_when crashes on type mismatch across branches
# Reproduction status: Fixed (dplyr 1.1+ coerces bare NA to dominant branch type)
# Type system catch: `number` rejects `number | null`

library(dplyr)

df <- data.frame(old = 1:3)

tryCatch({
  result <- df %>% mutate(new = case_when(
    old == 1 ~ 5,
    old == 2 ~ NA,
    TRUE ~ as.numeric(old)
  ))
  print(result)
}, error = function(e) {
  cat(paste("Crash:", e$message, "\n"))
})
